editor/plugins/node_3d_editor_plugin.cpp

```diff
@@ -7679,11 +7679,51 @@ Vector<int> Node3DEditor::get_subgizmo_selection() {
 }
 
 void Node3DEditor::add_control_to_menu_panel(Control *p_control) {
-	context_menu_hbox->add_child(p_control);
+	ERR_FAIL_NULL(p_control);
+	ERR_FAIL_COND(p_control->get_parent());
+
+	VSeparator *sep = memnew(VSeparator);
+	context_toolbar_hbox->add_child(sep);
+	context_toolbar_hbox->add_child(p_control);
+	context_toolbar_separators[p_control] = sep;
+
+	p_control->connect("visibility_changed", callable_mp(this, &Node3DEditor::_update_context_toolbar));
+
+	_update_context_toolbar();
 }
 
 void Node3DEditor::remove_control_from_menu_panel(Control *p_control) {
-	context_menu_hbox->remove_child(p_control);
+	ERR_FAIL_NULL(p_control);
+	ERR_FAIL_COND(p_control->get_parent() != context_toolbar_hbox);
+
+	p_control->disconnect("visibility_changed", callable_mp(this, &Node3DEditor::_update_context_toolbar));
+
+	context_toolbar_hbox->remove_child(context_toolbar_separators[p_control]);
+	context_toolbar_hbox->remove_child(p_control);
+	context_toolbar_separators.erase(p_control);
+
+	_update_context_toolbar();
+}
+
+void Node3DEditor::_update_context_toolbar() {
+	bool has_visible = false;
+	bool first_visible = false;
+
+	for (int i = 0; i < context_toolbar_hbox->get_child_count(); i++) {
+		Control *child = Object::cast_to<Control>(context_toolbar_hbox->get_child(i));
+		if (!child || !context_toolbar_separators.has(child)) {
```

Comment:
You could use the HashMap::Iterator trick here instead of `has()`, but I guess this code is less performance important.

Reply:
So it would look like this...

```cpp

		Control *child = Object::cast_to<Control>(context_toolbar_hbox->get_child(i));
		if (!child) {
			continue;
		}
		HashMap<Control *, VSeparator *>::Iterator E = context_toolbar_separators.find(child);
		if (!E) {
			continue;
		}

		if (child->is_visible()) {
			first_visible = !has_visible;
			has_visible = true;
		}
		E->value->set_visible(!first_visible && child->is_visible());
```

The size here should be pretty small, and not called too too often. So I think in this case a bit of performance hit for more clear and readable code would be better.